minetest/minetest
# Add particle blend mode "clip"

The first commit fixes the fact that apparently two lines of code were forgotten, so the modern `texture` properties didn't work in `add_particle`.

The second commit adds a "feature", which is really a special case fix for #3761: `blend = "clip"` allows particles which do not need alpha blending to be visible through translucent nodes, because Irrlicht renders solid scene nodes first (as it should).

I also see this as a soft prerequisite for correct batched particle rendering (#14155): Clipped particles can be safely batched (because they are rendered first). The same doesn't hold for blended particles.

## To do

This PR is Ready for Review.

I am considering additionally having clients default to "clip" if the texture doesn't make use of alpha.

## How to test

Backwards compatibility considerations:

* New client, old server: Nothing special should happen here.
* Old client, new server: This is slightly interesting. The old client should default to `blend = "alpha"`, ~~but we could also explicitly send this~~ https://github.com/minetest/minetest/pull/14353/commits/43b6be9a619c1447268d88b213d11bd312c560c6.

Spawn the "clip" test particle by right-clicking with the particlespawner testtool. Observe that it is visible through water:

![Screenshot from 2024-02-05 19-19-40](https://github.com/minetest/minetest/assets/34514239/7047ea82-1140-41bf-b6d4-47df73c8f08c)
